packages/opto-wallet/src/lib/opto-wallet.spec.ts

```diff
@@ -0,0 +1,109 @@
+/* eslint-disable @nrwl/nx/enforce-module-boundaries */
+import { Near, WalletConnection, ConnectedWalletAccount } from "near-api-js";
+import { AccountView } from "near-api-js/lib/providers/provider";
+import { mock } from "jest-mock-extended";
+
+import { mockWallet } from "../../../core/src/lib/testUtils";
+import type { MockWalletDependencies } from "../../../core/src/lib/testUtils";
+import type { BrowserWallet } from "../../../core/src/lib/wallet";
+
+const createOptoWallet = async (deps: MockWalletDependencies = {}) => {
+  const walletConnection = mock<WalletConnection>();
+  const account = mock<ConnectedWalletAccount>();
+
+  jest.mock("near-api-js", () => {
+    const module = jest.requireActual("near-api-js");
+    return {
+      ...module,
+      connect: jest.fn().mockResolvedValue(mock<Near>()),
+      WalletConnection: jest.fn().mockReturnValue(walletConnection),
+    };
+  });
+
+  walletConnection.isSignedIn.calledWith().mockReturnValue(true);
+  walletConnection.getAccountId
+    .calledWith()
+    .mockReturnValue("test-account.testnet");
+  walletConnection.account.calledWith().mockReturnValue(account);
+  // @ts-ignore
+  // near-api-js marks this method as protected.
+  // TODO: return value instead of null
+  account.signAndSendTransaction.calledWith().mockReturnValue(null);
+  account.state.calledWith().mockResolvedValue(
+    mock<AccountView>({
+      amount: "1000000000000000000000000",
+    })
+  );
+
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const { setupOptoWallet } = require("./opto-wallet");
+  const { wallet } = await mockWallet<BrowserWallet>(setupOptoWallet(), deps);
+
+  return {
+    nearApiJs: require("near-api-js"),
+    wallet,
+    walletConnection,
+    account,
+  };
+};
+
+afterEach(() => {
+  jest.resetModules();
+});
+
+describe("signIn", () => {
+  it("sign into near wallet", async () => {
+    const { wallet, nearApiJs } = await createOptoWallet();
+
+    await wallet.signIn({ contractId: "test.testnet" });
+
+    expect(nearApiJs.connect).toHaveBeenCalled();
+  });
+});
```

Comment:
All of the tests seem to be failing and I think they fail for the same reason:

When this code is executed:
```ts
const { wallet, nearApiJs } = await createOptoWallet();
```
It will eventually call the setup of Opto wallet `setupOptoWallet()` and in there you have an if check:
```ts
if (!window.opto) {
  return Promise.resolve(null);
}
```

So this means you will always get a `null` value for the `wallet`, and there's some code in the core lib when we try to access a module by `id` but since the value is `null` it will throw an error.

To fix this:

Add this at the top of the tests file below imports:

```ts
declare global {
  interface Window {
    opto: object | undefined;
  }
}
```
And then in your tests assign a value to `window.opto` so it does not return `undefined`, then an example test case would be something like this:


```ts
describe("signIn", () => {
  it("sign into near wallet", async () => {
    window.opto = {}; // note this is just added to make the test pass you can add a real value.
    const { wallet, nearApiJs } = await createOptoWallet();

    await wallet.signIn({ contractId: "test.testnet" });

    expect(nearApiJs.connect).toHaveBeenCalled();
  });
});
```